Create rule S6872 (#3542)
SonarSource/rspec: rules/S6872/cfamily/rule.adoc

Except for templated code, brace initialization should be preferred over {cpp}20's parenthesis initialization for aggregates.

== Why is this an issue?

With {cpp}20, it is now possible to initialize aggregate types using parentheses.

> This language feature was introduced to simplify writing generic code and consistently initialize objects, whether they are aggregates or not.

Reviewer: Like use of `generic` here. This could be as well referring to code expanded from macro, or generated by tool.

> For the sake of simplicity, aggregate types include arrays, unions, and structures without user-provided constructors and with only public non-static data members and public bases.

Reviewer: Could we have a note about this being different in different standard:
* C++14: default member initializers, bases
* C++17: deleted were ok

Reviewer: And reffer to the cpprefernece side.

> For the sake of simplicity, aggregate types include arrays, unions, and structures without user-provided constructors and with only public non-static data members and public bases.

Reviewer: ```suggestion
For the sake of simplicity, aggregate types include arrays, unions, and structures without user-declared constructors and with only public non-static data members and public bases.
```
This is current state.

> While parentheses are useful when initializing objects in templated code, they have several downsides compared to braces.

Reviewer: I do not thing that repiting that entry is ncessary.
```suggestion
Initializing objects with parentheses have several downsides compared to braces.
```

   For example, ``++Aggregate a(std::string());++`` declares function, while ``++Aggregate a{std::string()};++`` declares a variable.

 * Furthermore, using braces is idiomatic and consistent with C.

For all these reasons, braces should be preferred for non-generic code when initializing aggregates.
And the fix is often trivial: replace the parentheses `()` with braces `{}`.

Here is a noncompliant example:

[source,cpp,diff-id=1,diff-type=noncompliant]
----
struct Coordinate {
  int x;
  int y;
};

long readInteger();

auto readCoordinate() {
  // Be aware of the narrowing conversions on the next line.
  return Coordinate(readInteger(), readInteger()); // Noncompliant
}
----

There are multiple ways of handling the narrowing conversion; here is one alternative:

[source,cpp,diff-id=1,diff-type=compliant]
----
struct Coordinate {
  int x;
  int y;
};

long readInteger();

auto readCoordinate() {
  // Explicitly handle the conversion:
  // Here, we saturate, but throwing an exception may also be appropriate.
  auto readInt = []() { return saturate_cast<int>(readInteger()); }
  return Coordinate{readInt(), readInt()}; // Compliant
}
----

=== Exceptions

> Using `()` (i.e., with no arguments) to initialize an aggregate is accepted.
> The arguments listed above do not apply since the intent here is to use _value-initialization_.

Reviewer 2: ```suggestion
_Value-initialization_ with `()` (i.e., with no arguments) is unrelated to _aggregate-initialization_ and is accepted.
None of the drawbacks listed above apply in such situations.
```

> The arguments listed above do not apply since the intent here is to use _value-initialization_.

Reviewer: Would be good to say somthing was value_initializtion is, like saying it zero initializes trivial types and calls constructors on all subobjects.

// There are also edge cases not worth covering or even mentioning here.
//
// Example: having a member with an explicit default constructor does not compile.
// https://godbolt.org/z/exerMGM9x

[source,cpp]
----
Coordinate x(); // Irrelevant -- this is a function declaration.

auto x = Coordinate(); // Value-initialization - compliant by exception.

struct Object {
  Coordinate coord;
  Object() : coord() { } // Value-initialization - compliant by exception.
};
----

== Resources

=== Documentation

 * {cpp} reference -- https://en.cppreference.com/w/cpp/language/aggregate_initialization#Definitions[Aggregate definition]
 * {cpp} reference -- https://en.cppreference.com/w/cpp/language/value_initialization[Value-initialization]
 * Wikipedia -- https://en.wikipedia.org/wiki/Most_vexing_parse[Most vexing parse]

=== Standards

>  * Open Standards - https://open-std.org/JTC1/SC22/WG21/docs/papers/2015/n4462.html[N4462] presents the problem with initializing objects in generic code using braces, and https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p0960r2.html[P0960] solves it by introducing parenthesized initialization.

Reviewer 2: I did not link to https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p0960r0.html: yes, r0 has more examples than r2 but none of them are about templates so it may give the wrong idea to readers.

Reviewer: Makes sense.

Reviewer: I think that the solving paper is not needed. I would just say:
> in generic code using braces, that was solved by the introduction of parenthesized initialization.

Reviewer 2: Done in 9a4a3b120b

=== Related rules

 * S835 - Braces should be used to indicate and match the structure in the non-zero initialization of arrays and structures

> // TODO CPP-4792 - Update S835's title and list S6871.

Reviewer 2: To be taken care of before merging.
